Reregistration GR-PEACH as mbed support board

TomoYamanaka opened the pull request:
## Description

By #5342, since Cortex-A was supported wtih updating CMSIS/RTX, I request that **GR-PEACH** is registed as mbed board again.
Therefore, I PR the process that depend on GR-EPACH, and more.

## Request
- I would like to compile the Cortex-A boards that is supported as Mbed. 
  To do that, it's necessary to change the python file in "tools".
  In "tools/build_api.py", please remove the processing done by #4415.

- I strongly desire to be included in the shortest release package after completing the merge of this PR.

##
@toyowata 
Please help me to go ahead with the above.

0xc0170 commented:
> I would like to compile the Cortex-A boards that is supported as Mbed.
To do that, it's necessary to change the python file in "tools".
In "tools/build_api.py", please remove the processing done by #4415.

This should be part of this PR, or should be done as separate patch and this commited on top of it, to test these changes.
@theotherjimmy Can we just revert the PR 4415 as whole? Please advise

@TomoYamanaka Was this tested with the modifications (reverting locally 4415) ?

TomoYamanaka commented:
@0xc0170  

> @TomoYamanaka Was this tested with the modifications (reverting locally 4415) ?

Of course. I tested with modification it. :)

There was a lack in the above description. I needed to modify "tools\toolchains\arm.py" too.
https://github.com/ARMmbed/mbed-os/commit/b0fc103de6e1f93893ef1854f02933e54fc1b827#diff-9cf75aea879b6b9cc5026af0b8f708a3R37
I added **"Cortex-A"** in arm.py because I couldn't compile by ARMCC.

theotherjimmy commented:
@TomoYamanaka @0xc0170 That PR was designed to be reverted at a later date. Please feel free to revert.

TomoYamanaka commented:
@theotherjimmy 

> That PR was designed to be reverted at a later date. Please feel free to revert.

I reverted "Remove Cortex-A mbed OS 5 support".

Also, 
> There was a lack in the above description. I needed to modify "tools\toolchains\arm.py" too.

I added Cortex-A for core support in ARM toolchain.

0xc0170 commented on targets/TARGET_RENESAS/TARGET_RZ_A1H/us_ticker.c:
```diff
@@ -62,6 +62,7 @@ void us_ticker_init(void) {
     // INTC settings
     InterruptHandlerRegister(US_TICKER_TIMER_IRQn, (void (*)(uint32_t))us_ticker_interrupt);
     GIC_SetPriority(US_TICKER_TIMER_IRQn, 5);
+    GIC_SetConfiguration(US_TICKER_TIMER_IRQn, 3);
```
The commit msg:

> Change the processes that depend on GR-EPACH.
> - These files are based on the codes of CMSIS5/RTX5 for RZ/A1H incorporated in DS-5.
 > - I modified the support range from Mbed OS 2 to Mbed OS 2 and 5.
 > - In "mbed_rtx.h", I changed the values depend on GR-PEACH like a other targets for Cortex-M.

Is this change related to the first one `These files are based on the codes of CMSIS5/RTX5 for RZ/A1H incorporated in DS-5. ` ? What does this change do? Changes like this should be in a separate commit - what is fixing and how.

TomoYamanaka replied:
By the codes of CMSIS5/RTX5 for RZ/A1H incorporated in DS-5, `GIC_SetConfiguration` is a newly added function as Cortex-A, this function sets the interrupt configration using GIC.
It seems that this processing has become necessary for interrupt of each driver to work properly.

0xc0170 replied:
Thanks, that answers it. Please consider this type of details to be part of the change (using git commit message).

You can amend the commit msg to include there this detail for the reader, it would answer my question.

TomoYamanaka replied:
I amended the commit msg. Make a sense?

0xc0170 commented on rtos/TARGET_CORTEX/mbed_boot.c:
```diff
@@ -187,7 +187,11 @@ osThreadAttr_t _main_thread_attr;
 #define MBED_CONF_APP_MAIN_STACK_SIZE 4096
 #endif
+#if defined(__CORTEX_A9)
+MBED_ALIGN(8) char _main_stack[MBED_CONF_APP_MAIN_STACK_SIZE*4];
```
why we cant use the default one (*4 here)? This is breaking this config. As a user would set it to 10k for instance, actually would get 40k.

TomoYamanaka replied:
This change refer in the processing of previous mbed-os version.
Reference is [here](https://github.com/ARMmbed/mbed-os/blob/mbed-os-5.4.7/rtos/rtx/TARGET_CORTEX_A/RTX_CM_lib.h#L423).
I think Cortex-A in general will require more stack space more than Cortex-M.

0xc0170 replied:
As we can see, some changes like this would be good to have as a separate commit with description to understand  the reason for this change.

The default value should be rather changed. It is set on the line 188 in this file. As we would need to do *4 for any line that uses it in this file.

> I think Cortex-A in general will require more stack space more than Cortex-M.

4k by default would not be sufficient to run mbed-os examples ? If we agree that Cortex-A should have 4x more stack by default than cortex-m, then we would need:

```
#if cortex-a
// default stack is 16k
#else
// default stack is 4k
#endif
```

@bulislaw Please check

bulislaw replied:
We decided to set it to 4K by default for all platforms and make it configurable for applications. Overwriting it silently like that is confusing. Why do we need to do that? Is 4K not enought to run basic Mbed Apps and examples?

TomoYamanaka replied:
In case of [this test](https://github.com/ARMmbed/mbed-os/blob/master/TESTS/mbedmicro-rtos-mbed/malloc/main.cpp), Executing with `#define THREAD_STACK_SIZE   256` in Cortex-A, program will abort because of the lack of stack.
As a result of check, in Cortex-A, program will work properly by `#define THREAD_STACK_SIZE    512` at least.
(Now the value of macro I committed is [2048](https://github.com/ARMmbed/mbed-os/pull/5530/commits/aa328244abad0ec0d08b7bf5f0755551b705ec04#diff-03297e82aaace147c0ac9db765ba2780R27), I will change to 512)

This is why, Cortex-A needs more stack size than Cortex-M.
Since this value(16K) may be too large, how about change the value of this from 16K to 8K by following the above size?
If there are no problems with my propose, I will change my commit soon.
Not to mention the change of `THREAD_STACK_SIZE` in the above test.

TomoYamanaka replied:
@0xc0170 @bulislaw
I will appreciate if you reply this promptly.

0xc0170 replied:
I would leave the `MBED_CONF_APP_MAIN_STACK_SIZE` as it is for all platforms. The question above was not answered, if 4k is not sufficient for mbed OS app. Did you have any failures with it - tests or examples?
This is application specific, it is intended to be tuned by application.

TomoYamanaka replied:
@0xc0170 
Thank you for your commnets.
I understood the specification of `MBED_CONF_APP_MAIN_STACK_SIZE`.
To be honest, I was worried about the lack of the stack by any chance, there was no problem for mbed test in case of 4K. 
Therefore, I will modify to revert the value of macro to 4K soon.

TomoYamanaka replied:
I committed.

0xc0170 commented on targets/targets.json:
```diff
@@ -2356,11 +2356,12 @@
         "core": "Cortex-A9",
         "program_cycle_s": 2,
         "extra_labels": ["RENESAS", "MBRZA1H"],
+        "macros": ["TARGET_FEATURE_EXTENSION_REGISTER_COUNT_GCC=32", "TARGET_FEATURE_EXTENSION_REGISTER_COUNT_IAR=32"],
```
What does these 2 macro do? Would it be better to put into the config, instead of macro ?

I can see it is used in  the rtx assembly files, are these also part of RTX ?

TomoYamanaka replied:
Original was the form of `TARGET_FEATURE_EXTENSION_REGISTER_COUNT`.
However, since this macro is a built-in environment variable of the ARM compiler,
in the case of GCC or IAR, this macro is became un-active, either process related to VFP does not run.
The macro here should be the target dependent value. In order to run the processing with the register count 32 as same previously, 
I prepared the renamed macro, I actived it's processes in GCC and IAR.
Finally there was no a file is included in RTX assembly file, so I defined a macro in `targets.json`.

0xc0170 replied:
Can we do this via config, see an example: https://github.com/ARMmbed/mbed-os/blob/master/targets/targets.json#L269. Rather than macro with some value?

TomoYamanaka replied:
As a result of cherry-picking fixes from official CMSIS repo please, the macro that I added in targets.json became unnecessary. So I reverted the change of its file.

JonatanAntoni commented on rtos/TARGET_CORTEX/rtx5/RTX/Source/TOOLCHAIN_GCC/TARGET_CORTEX_A/irq_ca.S:
```diff
@@ -391,13 +391,13 @@ osRtxContextSave:
                 VMRS    R2, FPSCR
                 STMDB   R3!, {R2,R12}              // Push FPSCR, maintain 8-byte alignment
-                #if TARGET_FEATURE_EXTENSION_REGISTER_COUNT == 16
+                #if TARGET_FEATURE_EXTENSION_REGISTER_COUNT_GCC == 16
```
We fixed that issue in the meanwhile. Please check with latest CMSIS release if this works for you.

JonatanAntoni commented on rtos/TARGET_CORTEX/rtx5/RTX/Source/TOOLCHAIN_IAR/TARGET_CORTEX_A/irq_ca.S:
```diff
@@ -370,19 +369,19 @@ osRtxContextSave
+                ;ORR     R2, R2, #2
+                ;STRB    R2, [R0, #TCB_SP_FRAME]
+                ;ENDIF
+                ;IF {TARGET_FEATURE_EXTENSION_REGISTER_COUNT_IAR} == 32
```
We fixed this issue on latest CMSIS release.

JonatanAntoni commented on rtos/TARGET_CORTEX/mbed_rtx_idle.cpp:
```diff
@@ -24,7 +24,11 @@
+#include "rtx_core_ca.h"
+#else
 #include "rtx_core_cm.h"
+#endif
```
Why do you need to access this internal RTX header?
A user should never need to rely on RTX implementation details like this.

bulislaw replied:
I don't actually think it's needed.

TomoYamanaka replied:
It seems that it implemented by this commit (https://github.com/ARMmbed/mbed-os/commit/4b354f80cfbd819a951b9ac93ea7c781f3d5a77b#diff-415fd494349b5414064e0cc0d9f0d596).
In cortex-A, the build error occurs by including "rtx_core_cm.h". Therefore I implemented to include the header file for Cortex-A("rtx_core_ca.h").

JonatanAntoni replied:
Sure rtx_core_cm.h contains the core abstraction for Cortex-M, so this cannot be used for Cortex-A. I strongly recommend not to rely on internal stuff. Actually the include (formerly core_cm.h) was introduced in #e44d94f. Can you point out what you need that's defined in this header?

bulislaw replied:
It's not needed and can be deleted. I'll submit an PR to do it.

JonatanAntoni commented on cmsis/TARGET_CORTEX_A/core_ca.h:
```diff
@@ -1372,18 +1372,18 @@ __STATIC_INLINE void GIC_DistInit(void)
       //Disable the SPI interrupt
-      GIC_DisableIRQ(i);
+      GIC_DisableIRQ((IRQn_Type)i);
       if (i > 15) {
```
i starts at 32, thus this condition is always true.
We fixed GIC implementation in latest CMSIS release.

JonatanAntoni commented on cmsis/TARGET_CORTEX_A/cmsis_iccarm.h:
```diff
@@ -581,6 +603,13 @@ void __FPU_Enable(void)
 
 
+__STATIC_INLINE
+int __disable_irq_iar() {
```
Its not the intention of CMSIS to incorporate device and/or compiler specific functions. All functions (API) shall rather be implemented for all compilers in this case. It seems you are looking for a extended version of __disable_irq() that returns the value of CPSR before modification.

TomoYamanaka replied:
I reverted this function by cherry-picking from CMSIS repo, implemented disable_irq() for IAR by utilizing mbed critical function.

bulislaw commented:
@TomoYamanaka Instead of fixing it yourself, could you cherry pick fixes from official CMSIS repo please? All the other modifications to CMSIS should be submitted and accepted to CMSIS repo as well and then cherry picked on top of Mbed.

TomoYamanaka commented:
@bulislaw 
Now, I'm preparing to cherry-pick them, so can you wait a second?


@theotherjimmy 

> Could you add "Cortex-A9" to ARMC6's supported core list as well?

I commited this change.

TomoYamanaka commented:
I cherry-picked fixes from official CMSIS repo please.
And I modifid the macro of "THREAD_STACK_SIZE" using malloc test to 512B.

bulislaw commented:
I'm sorry that's not cherry picking: https://github.com/ARMmbed/mbed-os/pull/5530/commits/66aea98350b015445f74de47a18ec1bffaf83a32 could we have actual commits 1 to 1 from the CMSIS repo so we know which fixes we actually have.

bulislaw commented:
These commit still change CMSIS/RTX:
https://github.com/ARMmbed/mbed-os/pull/5530/commits/54a3e567c82b9c7eb91ca5f3af67159cd7b23640
https://github.com/ARMmbed/mbed-os/pull/5530/commits/21d0d311548c4e86192fb22c97c61c54900fd5c0

If they are not longer needed because of the 'cherry-picking' please remove them. If they are still needed can we submit them to CMSIS/RTX repo?

TomoYamanaka commented:
@bulislaw 
Thank you for your kindness.
I will try again.

TomoYamanaka commented:
@bulislaw

I cherry-picked from CMSIS repo.
Also I changed the method of disable_irq() / enable_irq() in each drivers.


> If they are not longer needed because of the 'cherry-picking' please remove them. If they are still needed can we submit them to CMSIS/RTX repo?

I adjusted the contents of CMSIS repo, but there were two changes needed to run GR-PEACH in cmsis_iccarm.h.
https://github.com/ARMmbed/mbed-os/pull/5530/commits/6239ec0747f04640f59694ecd771beacfe5d79d6
https://github.com/ARMmbed/mbed-os/pull/5530/commits/31ce59a41d12ab538c82af5af5e0e85dfd9dfc54

Could you sumit them to CMSIS repo ?

bulislaw commented:
@TomoYamanaka I'm sorry I can't see anything in this history, can you remove all the unnecessary commits rather than overwrite them and change back and forth? The log should be clean as we are reviewing commit by commit and also it'll end up in the Mbed OS history.

Also if your commits will be structured as follows it will build a coherent story and make all our lives easier.
1..a) Changes to PEACH code
a..b) Changes to Mbed OS including reverting disabling the Cortex A
b..c) Cherry-picked changes from CMSIS/RTX
c..d) New changes to CMSIS/RTX that @JonatanAntoni can review and then can be submitted to https://github.com/ARM-software/CMSIS_5/

TomoYamanaka commented:
I've closed This pull request by mistake, so I will PR again soon.
